### PR TITLE
Add a playback threshold percentage at which point to mark an episode as played

### LINKED ISSUE #106: FR/Discussion: Better logic to mark episode as played
Currently Doughnut marks the episode as played after 90% ( [`private static let playedThreshold: Double = 0.9`](https://github.com/dyerc/Doughnut/blob/main/Doughnut/Player/Player.swift#L38-L38) ), which if fine most of the time, especially for shorter/regular podcasts (30 minutes gives you 3 minutes left, which quite often is also just promotion). For longer episodes (more than 2h) it rises to almost 15 minutes, which quite often covers entire section of the podcast.

I'm not really sure what would be the best solution here. Maybe use fixed time (if left less than x minutes)? But that's more rigid and could break for shorter episodes (less than 5 minutes). Maybe increase the percentage to 95-97% of played time? Possibly make it configurable (advanced section, user could choose percentage or fixed remaining time and possibly set the value bye themselves)?

### PR DESCRIPTION
**Pull request by dyerc**
This PR fixes #106.

@GetToSet would you mind reviewing please. It seemed necessary to have an additional observer event for when the player finishes playback. Do you think we should reset the player visual state at this point? Currently it just stays the same but with 0:00 remaining on the clock (as it always has done). Nothing necessarily short-term, but it was something that occurred to me.

### CODE REVIEW

File: Doughnut/Player/Player.swift [hunk 1]
```diff
@@ -53,6 +51,18 @@ final class Player: NSObject {
 
   private(set) var pausedAt: TimeInterval? = nil
 
+  var playedThreshold: Double {
+    get {
```
Comment (GetToSet) [inline Doughnut/Player/Player.swift:55]: Swift code generally omits the `get` keyword for readonly property getters.

SwiftLint has [a default rule](https://realm.github.io/SwiftLint/implicit_getter.html) for this and actually I've planned to enable that in the next version.

**Comment by GetToSet** [comment]
> This PR fixes #106.
> 
> @GetToSet would you mind reviewing please. It seemed necessary to have an additional observer event for when the player finishes playback. Do you think we should reset the player visual state at this point? Currently it just stays the same but with 0:00 remaining on the clock (as it always has done). Nothing necessarily short-term, but it was something that occurred to me.

I think there's no need to cleanup player states when played to the end and we'll revisit the behavior when adding the playlist feature. Further, should we set the default threshold to 90% (to match the old behavior) or keep the default value 100%?

File: Doughnut/Player/Player.swift [hunk 2]
```diff
@@ -182,7 +192,7 @@ final class Player: NSObject {
           episode.playPosition = Int(time.seconds)
           episode.duration = Int(avPlayer.currentItem?.asset.duration.seconds ?? 0)
 
-          if episode.duration > 0, (Double(episode.playPosition) / Double(episode.duration)) > Self.playedThreshold {
+          if episode.duration > 0, (Double(episode.playPosition) / Double(episode.duration)) > self.playedThreshold {
```
Comment (GetToSet) [inline Doughnut/Player/Player.swift:195]: Update: seems `AVPlayer` always fires the timeObserver when played to the end and there's no rounding issues happening. Just changing `>` to `>=` would be enough.

**Comment by dyerc** [comment]
> > This PR fixes #106.
> > @GetToSet would you mind reviewing please. It seemed necessary to have an additional observer event for when the player finishes playback. Do you think we should reset the player visual state at this point? Currently it just stays the same but with 0:00 remaining on the clock (as it always has done). Nothing necessarily short-term, but it was something that occurred to me.
> 
> I think there's no need to cleanup player states when played to the end and we'll revisit the behavior when adding the playlist feature. Further, should we set the default threshold to 90% (to match the old behavior) or keep the default value 100%?

I think we can set the new value to 100% because it wasn't shown anywhere that the episode is marked as played after 90%, unless the user looked at the source code.

I have left in the `playerDidFinishPlaying` observer because I like the idea of resetting the play position of the episode to 0. If the user plays it again, starting from the beginning seems logical.

Thanks for your testing @GetToSet 👍

**Pull request merged**